Bug 1802534: gcp-routes: move to MCO, implement downfile, tweak timing

squeed opened the pull request:
Move gcp-routes to the MCO, call it openshift-gcp-routes. Also, improve the behavior significantly:

1. always accept connections from external sources
2. use downfiles instead of stopping the service
3. implement the equivalent in gcp-routes-controller.

squeed commented:
cc @sttts

sttts commented on cmd/gcp-routes-controller/README.md:
```diff
@@ -0,0 +1,23 @@
+
+## Background
+
+Google cloud load balancer is a L3LB that is Special. It doesn't do DNAT; instead, it
```
nit: special

squeed replied:
hah, needless capitalization is classic Nerd English but I'll change it :p

sttts commented on cmd/gcp-routes-controller/README.md:
```diff
@@ -0,0 +1,23 @@
+
+So, an agent exists on the node. It programs the node (either via iptables or routing tables) to
+accept traffic destined for the VIP. However, this has a problem: all hairpin traffic
+to the balanced servce is *always* handled by that backend.
```
... even if it is down or unhealthy.

squeed commented:
PR adding downfile support to gcp-routes.sh: https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899

sttts commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -82,15 +97,14 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-		FailureThreshold: 10,
-		OnFailure:        func() error { return exec.Command("systemctl", "stop", runOpts.gcpRoutesService).Run() },
-		OnSuccess:        func() error { return exec.Command("systemctl", "start", runOpts.gcpRoutesService).Run() },
+		FailureThreshold: 5, // it takes 20 seconds for vip changes to propagate, so give us 25 seconds before we react
```
would even go down to 3

squeed replied:
yeah, fair enough; it's just kubelet, sdn, and multus that use this.

sttts commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+	if h.mode == modeStopService && runOpts.rootMount != "" {
+		glog.Infof(`Calling chroot("%s")`, runOpts.rootMount)
+		if err := syscall.Chroot(runOpts.rootMount); err != nil {
+			return nil, fmt.Errorf("Unable to chroot to %s: %s", runOpts.rootMount, err)
```
nit: lower-case go errors

sttts commented on templates/master/00-master/gcp/files/etc-kubernetes-manifests-gcp-routes-controller.yaml:
```diff
@@ -15,7 +15,7 @@ contents:
         args:
         - "run"
-        - "--health-check-url=https://127.0.0.1:6443/readyz"
+        - "--health-check-url={{.Infra.Status.APIServerInternalURL}}/readyz"
```
just to be sure, is this written once or dynamically updated? Afaik in theory APIServerInternalURL could change?

squeed replied:
I don't think anything supports this changing. AFAICT the "status" is generated in the installer.

sttts replied:
ok. then it's fine.

squeed commented:
/hold
until the gcp PR is merged and "deployed" in to CI.

LorbusChris commented:
/cherry-pick fcos
/cc @vrutkovs

squeed commented:
Seems the unit test scaffolding is out-of-date, so this is not going to pass unit tests. Investigating.

squeed commented:
Seems the unit test scaffolding is out-of-date, so this is not going to pass unit tests. Investigating.

umohnani8 commented:
/test e2e-gcp-op

tedyu commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -111,11 +125,10 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+			if err := handler.onFailure(); err != nil {
+				glog.Infof("Failed to mark service down on signal: %s", err)
 			}
+			os.Exit(0)
```
Should this be in the else block (err == nil) ?

squeed replied:
I don't think so - this is in the SIGINT handler. Kubelet is asking us to shut down and go away.

tedyu commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+		} else {
+			addrs, err := net.LookupHost(uri.Hostname())
+			if err != nil {
+				return nil, fmt.Errorf("failed to lookup host: %v", err)
```
uri.Hostname() should be included in the message

tedyu commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+				return nil, fmt.Errorf("failed to lookup host: %v", err)
+			}
+			if len(addrs) != 1 {
+				return nil, fmt.Errorf("hostname %s has %d addresses, expected 1 - aborting", uri.Hostname(), len(addrs))
```
Looking at line 185 below, we are not aborting

squeed replied:
Not sure what you mean - we return err and handle that on line 81?

tedyu replied:
If we abort, is the "Using VIP %s" log necessary ?

tedyu commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -129,6 +142,90 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
+			return fmt.Errorf("failed to create downfile (%s): %v", downFile, err)
+		}
+		_ = fp.Close()
+		glog.Infof("healthcheck failed, created %s", downFile)
```
It would make log searching easier if 'downFile' is included in the message (as is done on line 198 above)

squeed replied:
good point, updated.

kikisdeliveryservice commented:
seeing:
`E0421 19:17:08.844158       1 leaderelection.go:331] error retrieving resource lock openshift-sdn/openshift-network-controller: Get https://api-int.ci-op-jd83nzz1-1354f.origin-ci-int-gce.dev.openshift.com:6443/api/v1/namespaces/openshift-sdn/configmaps/openshift-network-controller: dial tcp 10.0.0.2:6443: i/o timeout`

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/1670/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1940/artifacts/e2e-gcp-op/pods/openshift-sdn_sdn-controller-2t46q_sdn-controller.log

squeed commented:
> seeing:
> `E0421 19:17:08.844158 1 leaderelection.go:331] error retrieving resource lock openshift-sdn/openshift-network-controller: Get https://api-int.ci-op-jd83nzz1-1354f.origin-ci-int-gce.dev.openshift.com:6443/api/v1/namespaces/openshift-sdn/configmaps/openshift-network-controller: dial tcp 10.0.0.2:6443: i/o timeout`
> 
> https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/1670/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1940/artifacts/e2e-gcp-op/pods/openshift-sdn_sdn-controller-2t46q_sdn-controller.log

Nuts. Investigating.

squeed commented:
Those error messages are red herrings; the node maintained access to the internal api server at that time. I suspect they're just from masters rebooting. Continuing the investigation.

squeed commented:
Okay, this is weird: tests are failing because one of the worker nodes failed to start the MCD after a reboot - because CRIO won't create the sandbox.

`Apr 21 21:21:32.525864 ci-op-d4rtc-w-b-xrmnd.c.openshift-gce-devel-ci.internal hyperkube[1630]: I0421 21:21:32.525753    1630 event.go:278] Event(v1.ObjectReference{Kind:"Pod", Namespace:"openshift-machine-config-operator", Name:"machine-config-daemon-4kr98", UID:"fb999c1d-c1ac-4e8f-bf2e-98e2b785acf4", APIVersion:"v1", ResourceVersion:"27971", FieldPath:""}): type: 'Warning' reason: 'FailedCreatePodSandBox' Failed to create pod sandbox: rpc error: code = Unknown desc = reserving pod sandbox name: error reserving pod name k8s_machine-config-daemon-4kr98_openshift-machine-config-operator_fb999c1d-c1ac-4e8f-bf2e-98e2b785acf4_1 for id 6bfb87bd099b9954b3ee5654c35e9b6fc8403df3f06f6f1a57331195f94dc28e: name is reserved`

This definitely isn't caused by my change - mine only touches the master nodes. But it's not good. I wonder if the problem lies with kubelet, crio, or libpod?

squeed commented:
Known (and very scary) bug: 1785399
/retest

squeed commented:
Updated, fixed unit test scaffolding. This job is pretty prone to triggering the crio leak, so it might be a while before this goes green. AFAICT it's not a bug in the PR.

knobunc commented:
/retest

mrunalp commented:
@squeed opened https://bugzilla.redhat.com/show_bug.cgi?id=1826896 to fix the issue you ran into.

kikisdeliveryservice commented:
> /hold
> until the gcp PR is merged and "deployed" in to CI.

@squeed has this other PR merged yet?

sttts commented:
> @squeed has this other PR merged yet?

It has: https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899

/lgtm
/approve

sttts commented:
/hold

checking whether https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/899 is in promoted image

runcom commented:
/approve

sttts commented:
For the record, don't merge this. With the current gcp-routes implementation it makes the situation worse.

squeed commented:
@sttts fixed the issues with the iptables scaffolding. When https://gitlab.cee.redhat.com/coreos/redhat-coreos/merge_requests/905 merges (and builds in to CI), then this should be good to go.

sttts commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -82,15 +97,14 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-		FailureThreshold: 10,
-		OnFailure:        func() error { return exec.Command("systemctl", "stop", runOpts.gcpRoutesService).Run() },
-		OnSuccess:        func() error { return exec.Command("systemctl", "start", runOpts.gcpRoutesService).Run() },
+		FailureThreshold: 3, // it takes at most 20 seconds for vip changes to propagate
```
please add more comments here why these values are crucial.

squeed replied:
done, and poked timings.

squeed commented on cmd/gcp-routes-controller/run.go:
```diff
@@ -82,22 +97,27 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
 	tracker := &healthTracker{
 		state:            unknownTrackerState,
 		ErrCh:            errCh,
 		SuccessThreshold: 2,
```
Is there any chance the apiserver can flap on startup? If not, then 1 is good.

sttts replied:
I think when it is ready it is ready. No reason to expect flapping.

squeed commented:
Okay, I've updated this to create an openshift-specific service and move it in to the MCO.

LorbusChris commented:
If we want to pull this out of the base OS and into MCO, I still wonder how are we going to run this on the bootstrap node?

squeed commented:
> If we want to pull this out of the base OS and into MCO, I still wonder how are we going to run this on the bootstrap node?

We'll have to keep a generic version in the base OS, without all the openshift customizations. Eventually that can move to nm-cloud-setup.

vrutkovs commented:
>I still wonder how are we going to run this on the bootstrap node?

bootstrap node could extract this file from MCO image? 

Perhaps its worth investing into `machine-config-daemon prepare-boostrap-node` subcommand which would set GCP routes file, update CRI-O config etc.

squeed commented:
Whelp, I can't get the changes to actually create a bootable master. And without the serial console logs, I'm not sure what the issue is.

LorbusChris commented:
There are no logs from the control-planes machines at all. This looks weird:
from `log-bundle-20200513194307/unit-status/afterburn-sshkeys@core.service.txt`:
```
● afterburn-sshkeys@core.service - Afterburn (SSH Keys)
   Loaded: loaded (/usr/lib/systemd/system/afterburn-sshkeys@.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Wed 2020-05-13 18:57:56 UTC; 45min ago
  Process: 1658 ExecStart=/usr/bin/afterburn ${AFTERBURN_OPT_PROVIDER} --ssh-keys=core (code=exited, status=1/FAILURE)
 Main PID: 1658 (code=exited, status=1/FAILURE)

May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.513 INFO Fetching http://metadata.google.internal/computeMetadata/v1/project/attributes/sshKeys: Attempt #1
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.514 INFO Fetch successful
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.515 INFO Fetching http://metadata.google.internal/computeMetadata/v1/project/attributes/ssh-keys: Attempt #1
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: May 13 18:57:56.518 INFO Fetch successful
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Error: writing ssh keys
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Caused by: invalid key format
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal afterburn[1658]: Caused by: Invalid last symbol 86, offset 366.
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: afterburn-sshkeys@core.service: Main process exited, code=exited, status=1/FAILURE
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: afterburn-sshkeys@core.service: Failed with result 'exit-code'.
May 13 18:57:56 ci-op-hxlbd-b.c.openshift-gce-devel-ci.internal systemd[1]: Failed to start Afterburn (SSH Keys).
```
and in `.openshift_install.log`:
```
time="2020-05-13T19:43:20Z" level=debug msg="Gather remote logs"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.5"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.5 port 22: Connection refused\r"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.3"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.3 port 22: Connection refused\r"
time="2020-05-13T19:43:20Z" level=debug msg="Collecting info from 10.0.0.4"
time="2020-05-13T19:43:20Z" level=debug msg="lost connection"
time="2020-05-13T19:43:20Z" level=debug msg="ssh: connect to host 10.0.0.4 port 22: Connection refused\r"
```

LorbusChris commented:
maybe this was a flake?^ Other e2e tests run fine.
/test e2e-gcp-op

squeed commented:
/test e2e-gcp-op
so I can grab some proposed ignition config.

squeed commented:
Aha!

`CRITICAL : Ignition failed: failed to create files: failed to create files: failed to create file "/usr/libexec/openshift-gcp-routes.sh": unlinkat /sysroot/usr/libexec/openshift-gcp-routes.sh: read-only file system`

squeed commented:
Jumped on to a node, saw that everything is working as planned - rhcos-provided service is disabled, new service up and running, iptables looks correct.

sttts commented:
/hold cancel

squeed commented:
Tweaked timing based on sttts' suggestion; updated.

sttts commented:
/retest

squeed commented:
those aren't real failures, they're cancelled jobs because of the force-push.

squeed commented:
e2e-gcp-op passed. I'm satisfied with this now. This is ready for final review.

ashcrow commented:
cc @runcom

kikisdeliveryservice commented:
@runcom has already approved this PR

deferring to @sttts for final LGTM

kikisdeliveryservice commented:
/test e2e-ovirt

ashcrow commented:
> @runcom has already approved this PR
> 
> deferring to @sttts for final LGTM

Ahh it was hidden in the 25 rolled up messages.

sttts commented:
/lgtm

squeed commented:
This seems to be a horrid test issue.
/hold
to stop the retests until it clears up

squeed commented:
/test e2e-aws

squeed commented:
/test e2e-aws

squeed commented:
/hold cancel
now that quay is back
/retest

squeed commented:
It looks like whatever the problem in https://bugzilla.redhat.com/show_bug.cgi?id=1828606 is biting us.

Given that this doesn't touch aws at all, would this be a candidate for an /override ?

sttts commented:
/cherry-pick release-4.4
